### ISSUE TITLE
Uncontrollable memory growth when growing the workspace during training/decoding

### ISSUE THREAD

**Issue by XapaJIaMnu** [marian-nmt/marian-dev#828]
### Bug description
In cases where the workspace is too small for a mini-batch during training  (when not using mini-batch-fit) or decoding (when there is unusually long sentence inside the mini-batch), the marian workspace [reallocs](https://github.com/marian-nmt/marian-dev/blob/db771e09bd90b839efa580bca62a6b08500a1f33/src/tensors/cpu/device.cpp#L17) more memory. This is done through the [grow](https://github.com/marian-nmt/marian-dev/blob/db771e09bd90b839efa580bca62a6b08500a1f33/src/tensors/allocator.h#L127) function. 

However, it seems that as this function is called multiple times, there is an issue of uncontrollable memory growth, which often results in an OOM crash. If the workspace is big enough so that grow is seldom called there is no uncontrollable memory growth.

Reproducible example can be found here: https://nbogoychev.com/files/memory_leak_example.tar.xz  . When training (on the CPU) with `mini-batch-fit: false` and `workspace: 2000` (config file provided in the archive: `config.debuglower.yml`), we get uncontrollable memory growth at the 11th mini-batch. Marian starts allocating more and more workspace memory, however the memory consumption grows much faster than the workspace. At some point marian reaches 770 GBs of memory usage, while the workspace sits at just 18 GBs. Eventually the system runs out of memory and kills marian.

In contrast if we increase the workspace to 40 GBs (config file provided in the archive: `config.debuglower40g.yml`), from the beginning, we have no extra allocations and no uncontrollable memory growth.

This is not a memory leak, as the memory is still reachable and when marian exits cleanly it deallocates it. Valgrind doesn't detect anything abnormal

I believe the issue is due to an STL container continuously growing and/or some unfortunate smart pointer circular references, which gets triggered as a by-process of the workspace expansion. I suspect https://github.com/marian-nmt/marian-dev/blob/db771e09bd90b839efa580bca62a6b08500a1f33/src/tensors/allocator.h#L91 as it increases its size by 1 after each realloc, but I don't have a proof. I believe this issue has existed in the codebase for a number of years and various people have run into this issue during decoding and the solution has always been "Increase the workspace" or "decrease the mini-batch size".

### How to reproduce
Run CPU training with the configurations provide here:  https://nbogoychev.com/files/memory_leak_example.tar.xz

### Context
* Marian version: Git master: `v1.10.3 6810afae 2021-03-02 08:42:55 +0000`
* CMake command: `cmake .. -DCOMPILE_CUDA=OFF`

**Comment by emjotde** [comment]
I assigned yourself to the bug since you seem to already spent more time on it than I am ready to invest at the moment. I assume this is also reproducible on CPU?

**Comment by XapaJIaMnu** [comment]
Yes, this happens both on CPU and GPU. The memory leaked is only ever in CPU space however, regardless of whether you're using a CPU or GPU to decode/train.

**Comment by XapaJIaMnu** [comment]
Ok, that took a while, but this is actually a old tcmalloc issue that is fixed in master. https://github.com/marian-nmt/marian-dev/blob/bb92b817dd5140523af9e4d756a34c2092d03e32/src/tensors/cpu/aligned.h#L36 this bit seems to have no effect when using tcmalloc 2.4 (ubuntu 16.04) or tcmalloc 2.5 (ubuntu 18.04). 

It does however work fine with stdlib malloc on 18.04 or with tcmalloc 2.9. I propose that we add tcmalloc to our 3rd party compiles and link statically to that one if the system tcmalloc is outdated? We can also put a warning in the CMakeLists.txt about the potential memory growth issue. What do you think?

**Comment by emjotde** [comment]
Oh. Good find! I would have never expected that.
We should check if tcmalloc is actually still doing anything for us. I think it was way more efficient before I introduced the custom smart pointers. Maybe we should just get rid of it? Do you have benchmarks for your stuff?

**Comment by XapaJIaMnu** [comment]
Normally TCMalloc performs faster than malloc in all cases at the cost of small memory overhead, especially with multiple small objects allocations (we still do have some stl allocations). I think we should still keep it.

That being said, I haven't done any benchmarks of tcmalloc vs no tcmalloc, but now that I know the issue I can rerun some of our CPU and GPU benchmarks.

**Comment by emjotde** [comment]
Would be much obliged. BTW. are the benchmarks public?

**Comment by XapaJIaMnu** [comment]
They are, but the scripts are not against a public branch. It's easy enough to change them. Should we add a `marian-benchmarks` subrepo?

**Comment by emjotde** [comment]
@snukky and I actually had plans to do exactly that. And I thing we even created the repo? 
Ah, it even has content: https://github.com/marian-nmt/marian-benchmarks :)

So yes, we should add your benchmarks there.

**Comment by XapaJIaMnu** [comment]
I will prepare and issue a pull.

**Comment by XapaJIaMnu** [comment]
Benchmark results (CPU) from my marian-benchmarks pull (1 million lines, en-de tiny wngt2020 model):
| type | time (tcmalloc) | time (stdlibmalloc) |
| ---    | --- | ---|
| intgemm8 | 420.05 | 418.48 |
| fbgemm8_avx512 | 391.00 | 394.19 |
| fp32    | 492.89 | 496.28  |

tl;dr seems to be fairly similar.

The intgemm version here doesn't include the output layer and precomputed activations quantisation multipliers that we had in the wngt submission.

**Comment by emjotde** [comment]
This is multi-thread?

**Comment by XapaJIaMnu** [comment]
N=32

**Comment by emjotde** [comment]
Great, in that case I would just disable TCMalloc unless 2.9 is available? Potentially output a message? Or should we just get rid of it?

**Comment by XapaJIaMnu** [comment]
Disable tcmalloc unless 2.9 is available and output a message? Tcmalloc is also beneficial for profiling (although anyone doing profiling should know how to link to the full tcmalloc).

**Comment by emjotde** [comment]
Re-opening as a reminder for myself. Removing TC_MALLOC has performance penalties after all.
